tests/test_response_required_logic.py

```diff
@@ -32,47 +32,50 @@ class FakeProductResponse:
     data: Dict[str, Any]
 
 
-@provides(DummyProductResponse)
 class DummyProductProvider(PageObjectInputProvider):
 
-    def __init__(self, request: scrapy.Request):
-        self.request = request
+    provided_classes = {DummyProductResponse}
 
-    def __call__(self):
+    def __call__(self, to_provide, request: scrapy.Request):
         data = {
             'product': {
-                'url': self.request.url,
+                'url': request.url,
                 'name': 'Sample',
             },
         }
-        return DummyProductResponse(data=data)
+        return {
+            DummyProductResponse: DummyProductResponse(data=data)
+        }
```

Comment:
```suggestion
        return [DummyProductResponse(data=data)]
```

Why weren't tests failing here?

Reply:
Because in this test the injection part is not tested, so this is not invoked. But It is not nice to have some code that is never called, so I'll add something to the tests. Sorry for missing all these changes and thank you for sporting them!